nni/compression/pytorch/speedup/compress_modules.py

```diff
@@ -3,196 +3,381 @@
-    'Sigmoid': lambda module, mask: no_replace(module, mask),
-    'Linear': lambda module, mask: replace_linear(module, mask),
-    'Dropout': lambda module, mask: no_replace(module, mask),
-    'Dropout2d': lambda module, mask: no_replace(module, mask),
-    'Dropout3d': lambda module, mask: no_replace(module, mask)
+    'BatchNorm2d': lambda module, auto_infer: replace_batchnorm2d(module, auto_infer),
+    'BatchNorm1d': lambda module, auto_infer: replace_batchnorm1d(module, auto_infer),
+    'Conv2d': lambda module, auto_infer: replace_conv2d(module, auto_infer),
+    'Linear': lambda module, auto_infer: replace_linear(module, auto_infer),
+    'MaxPool2d': lambda module, auto_infer: no_replace(module, auto_infer),
+    'AvgPool2d': lambda module, auto_infer: no_replace(module, auto_infer),
+    'AdaptiveAvgPool2d': lambda module, auto_infer: no_replace(module, auto_infer),
+    'ReLU': lambda module, auto_infer: no_replace(module, auto_infer),
+    'ReLU6': lambda module, auto_infer: no_replace(module, auto_infer),
+    'Dropout': lambda module, auto_infer: no_replace(module, auto_infer),
+    'Dropout2d': lambda module, auto_infer: no_replace(module, auto_infer),
+    'Dropout3d': lambda module, auto_infer: no_replace(module, auto_infer),
+    'LayerNorm': lambda module, auto_infer: replace_layernorm(module, auto_infer),
+    'ConvTranspose2d': lambda module, auto_infer: replace_convtranspose2d(module, auto_infer)
 }
 
+# NEED_FOLD_BIAS = True
 
-def no_replace(module, mask):
+
+def convert_to_coarse_mask(t_mask, dim):
+    """
+    Convert the mask tensor to the coarse-grained mask tensor.
+    Parameters
+    ---------
+    t_mask: torch.Tensor
+        The tensor only have 1s and 0s, 0 indicates this value is masked
+        and 1 indicates the corresponding value is not masked.
+    dim: int
+        Try to reduce the mask tensor on this dimension.
+
+    Returns
+    -------
+    indexes: torch.Tensor
+        The indexes of the sparsity that can be structurally removed.
+    remained_indexes: torch.Tensor
+        The indexes of values that need to be remained.
+    """
+    assert isinstance(t_mask, torch.Tensor)
+    shape = list(t_mask.size())
+    n_dims = len(shape)
+    dim_list = list(range(n_dims))
+    # try to reduce the mask from the dim-th dimension
+    dim_list.remove(dim)
+
+    t_merged = torch.sum(t_mask, dim_list)
+    assert t_merged.size(0) == shape[dim]
+    all_pruned = t_merged == 0
+    need_remain = t_merged != 0
+    # return the indexes of the sparsity that can be removed
+    indexes = torch.nonzero(all_pruned, as_tuple=True)[0]
+    remained_indexes = torch.nonzero(need_remain, as_tuple=True)[0]
+    return indexes, remained_indexes
+
+
+def no_replace(module, auto_infer):
     """
     No need to replace
     """
     _logger.debug("no need to replace")
     return module
 
 
-def replace_linear(linear, mask):
+def replace_linear(linear, auto_infer):
```

Comment:
it would be better to briefly describe what types of sparsity this function supports. in high-level, coarse-grained sparsity

Reply:
added